### ISSUE TITLE
text-generation-inference (TGI) support

### ISSUE THREAD

**Issue by sadransh** [pydantic/pydantic-ai#398]
I am having following issue with TGI self-hosted model when trying to use tool calling with the following code, while it is fine with  `result_type=str`

```python
from pydantic import BaseModel

from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIModel

class CityLocation(BaseModel):
    city: str
    country: str

model = OpenAIModel(
    model_name='Meta-Llama-3.1-70B-Instruct',
    base_url = 'http://localhost:8000/v1',
    api_key ="-",
    
)
agent = Agent(model, result_type=CityLocation)
result = agent.run_sync('Where were the olympics held in 2012?')
print(result.data)
#> city='London' country='United Kingdom'
print(result.cost())
```
error:
```
UnprocessableEntityError: Failed to deserialize the JSON body into the target type: messages[1]: missing field `content` at line 1 column 235
```
To my understanding based on tgi docs, it is openai compatible. 
https://huggingface.co/docs/text-generation-inference/en/messages_api
Does it require similar development to this feature request? https://github.com/pydantic/pydantic-ai/issues/224

environment: 

pydantic_ai v 0.0.13
python 3.9
pydantic 2.10.3

**Comment by samuelcolvin** [comment]
I've spent 30 minutes trying to run TGI on my macbook, and failed.

However I don't get this error with either OpenAI or ollama, so I think this must be an issue with TGI.

**Comment by sadransh** [comment]
Here is a quick way to try run TGI image on my M2 Macbook with small 135M param model: 

```
docker run  --platform linux/x86_64 --shm-size 1g -p 8000:80 -v $PWD/data:/data ghcr.io/huggingface/text-generation-inference:3.0.0 --model-id HuggingFaceTB/SmolLM2-135M-Instruct
```

`took 1min to fully load as its on 'cpu'`

testing if it is running fine: 
```
curl localhost:8000/info
```

then it is possible to reproduce the error by running the code from from. 

Here is a full error trackback: 

```
---------------------------------------------------------------------------
UnprocessableEntityError                  Traceback (most recent call last)
Cell In[1], line 19
     12 model = OpenAIModel(
     13     model_name='Meta-Llama-3.1-70B-Instruct',
     14     base_url = 'http://localhost:8000/v1',
     15     api_key ="-",
     16     
     17 )
     18 agent = Agent(model, result_type=CityLocation)
---> 19 result = agent.run_sync('Where were the olympics held in 2012?')
     20 print(result.data)
     21 #> city='London' country='United Kingdom'

File ~/miniconda3/envs/env/lib/python3.9/site-packages/pydantic_ai/agent.py:319, in Agent.run_sync(self, user_prompt, message_history, model, deps, model_settings, infer_name)
    317 if infer_name and self.name is None:
    318     self._infer_name(inspect.currentframe())
--> 319 return asyncio.get_event_loop().run_until_complete(
    320     self.run(
    321         user_prompt,
    322         message_history=message_history,
    323         model=model,
    324         deps=deps,
    325         infer_name=False,
    326         model_settings=model_settings,
    327     )
    328 )

File ~/miniconda3/envs/env/lib/python3.9/site-packages/nest_asyncio.py:90, in _patch_loop.<locals>.run_until_complete(self, future)
     87 if not f.done():
     88     raise RuntimeError(
     89         'Event loop stopped before Future completed.')
---> 90 return f.result()

File ~/miniconda3/envs/env/lib/python3.9/asyncio/futures.py:201, in Future.result(self)
    199 self.__log_traceback = False
    200 if self._exception is not None:
--> 201     raise self._exception
    202 return self._result

File ~/miniconda3/envs/env/lib/python3.9/asyncio/tasks.py:256, in Task.__step(***failed resolving arguments***)
    252 try:
    253     if exc is None:
    254         # We use the `send` method directly, because coroutines
    255         # don't have `__iter__` and `__next__` methods.
--> 256         result = coro.send(None)
    257     else:
    258         result = coro.throw(exc)

File ~/miniconda3/envs/env/lib/python3.9/site-packages/pydantic_ai/agent.py:251, in Agent.run(self, user_prompt, message_history, model, deps, model_settings, infer_name)
    248     agent_model = await self._prepare_model(model_used, deps, messages)
    250 with _logfire.span('model request', run_step=run_step) as model_req_span:
--> 251     model_response, request_cost = await agent_model.request(messages, model_settings)
    252     model_req_span.set_attribute('response', model_response)
    253     model_req_span.set_attribute('cost', request_cost)

File ~/miniconda3/envs/env/lib/python3.9/site-packages/pydantic_ai/models/openai.py:151, in OpenAIAgentModel.request(self, messages, model_settings)
    148 async def request(
    149     self, messages: list[ModelMessage], model_settings: ModelSettings | None
    150 ) -> tuple[ModelResponse, result.Cost]:
--> 151     response = await self._completions_create(messages, False, model_settings)
    152     return self._process_response(response), _map_cost(response)

File ~/miniconda3/envs/env/lib/python3.9/site-packages/pydantic_ai/models/openai.py:189, in OpenAIAgentModel._completions_create(self, messages, stream, model_settings)
    185 openai_messages = list(chain(*(self._map_message(m) for m in messages)))
    187 model_settings = model_settings or {}
--> 189 return await self.client.chat.completions.create(
    190     model=self.model_name,
    191     messages=openai_messages,
    192     n=1,
    193     parallel_tool_calls=True if self.tools else NOT_GIVEN,
    194     tools=self.tools or NOT_GIVEN,
    195     tool_choice=tool_choice or NOT_GIVEN,
    196     stream=stream,
    197     stream_options={'include_usage': True} if stream else NOT_GIVEN,
    198     max_tokens=model_settings.get('max_tokens', NOT_GIVEN),
    199     temperature=model_settings.get('temperature', NOT_GIVEN),
    200     top_p=model_settings.get('top_p', NOT_GIVEN),
    201     timeout=model_settings.get('timeout', NOT_GIVEN),
    202 )

File ~/miniconda3/envs/env/lib/python3.9/site-packages/openai/resources/chat/completions.py:1720, in AsyncCompletions.create(self, messages, model, audio, frequency_penalty, function_call, functions, logit_bias, logprobs, max_completion_tokens, max_tokens, metadata, modalities, n, parallel_tool_calls, prediction, presence_penalty, reasoning_effort, response_format, seed, service_tier, stop, store, stream, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)
   1678 @required_args(["messages", "model"], ["messages", "model", "stream"])
   1679 async def create(
   1680     self,
   (...)
   1717     timeout: float | httpx.Timeout | None | NotGiven = NOT_GIVEN,
   1718 ) -> ChatCompletion | AsyncStream[ChatCompletionChunk]:
   1719     validate_response_format(response_format)
-> 1720     return await self._post(
   1721         "/chat/completions",
   1722         body=await async_maybe_transform(
   1723             {
   1724                 "messages": messages,
   1725                 "model": model,
   1726                 "audio": audio,
   1727                 "frequency_penalty": frequency_penalty,
   1728                 "function_call": function_call,
   1729                 "functions": functions,
   1730                 "logit_bias": logit_bias,
   1731                 "logprobs": logprobs,
   1732                 "max_completion_tokens": max_completion_tokens,
   1733                 "max_tokens": max_tokens,
   1734                 "metadata": metadata,
   1735                 "modalities": modalities,
   1736                 "n": n,
   1737                 "parallel_tool_calls": parallel_tool_calls,
   1738                 "prediction": prediction,
   1739                 "presence_penalty": presence_penalty,
   1740                 "reasoning_effort": reasoning_effort,
   1741                 "response_format": response_format,
   1742                 "seed": seed,
   1743                 "service_tier": service_tier,
   1744                 "stop": stop,
   1745                 "store": store,
   1746                 "stream": stream,
   1747                 "stream_options": stream_options,
   1748                 "temperature": temperature,
   1749                 "tool_choice": tool_choice,
   1750                 "tools": tools,
   1751                 "top_logprobs": top_logprobs,
   1752                 "top_p": top_p,
   1753                 "user": user,
   1754             },
   1755             completion_create_params.CompletionCreateParams,
   1756         ),
   1757         options=make_request_options(
   1758             extra_headers=extra_headers, extra_query=extra_query, extra_body=extra_body, timeout=timeout
   1759         ),
   1760         cast_to=ChatCompletion,
   1761         stream=stream or False,
   1762         stream_cls=AsyncStream[ChatCompletionChunk],
   1763     )

File ~/miniconda3/envs/env/lib/python3.9/site-packages/openai/_base_client.py:1843, in AsyncAPIClient.post(self, path, cast_to, body, files, options, stream, stream_cls)
   1829 async def post(
   1830     self,
   1831     path: str,
   (...)
   1838     stream_cls: type[_AsyncStreamT] | None = None,
   1839 ) -> ResponseT | _AsyncStreamT:
   1840     opts = FinalRequestOptions.construct(
   1841         method="post", url=path, json_data=body, files=await async_to_httpx_files(files), **options
   1842     )
-> 1843     return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)

File ~/miniconda3/envs/env/lib/python3.9/site-packages/openai/_base_client.py:1537, in AsyncAPIClient.request(self, cast_to, options, stream, stream_cls, remaining_retries)
   1534 else:
   1535     retries_taken = 0
-> 1537 return await self._request(
   1538     cast_to=cast_to,
   1539     options=options,
   1540     stream=stream,
   1541     stream_cls=stream_cls,
   1542     retries_taken=retries_taken,
   1543 )

File ~/miniconda3/envs/env/lib/python3.9/site-packages/openai/_base_client.py:1638, in AsyncAPIClient._request(self, cast_to, options, stream, stream_cls, retries_taken)
   1635         await err.response.aread()
   1637     log.debug("Re-raising status error")
-> 1638     raise self._make_status_error_from_response(err.response) from None
   1640 return await self._process_response(
   1641     cast_to=cast_to,
   1642     options=options,
   (...)
   1646     retries_taken=retries_taken,
   1647 )

UnprocessableEntityError: Failed to deserialize the JSON body into the target type: messages[1]: missing field `content` at line 1 column 248
```

and here is `agent.last_run_messages`
```
[ModelRequest(parts=[UserPromptPart(content='Where were the olympics held in 2012?', timestamp=datetime.datetime(2024, 12, 18, 18, 52, 59, 258702, tzinfo=datetime.timezone.utc), part_kind='user-prompt')], kind='request'),
 ModelResponse(parts=[ToolCallPart(tool_name='final_result', args=ArgsJson(args_json={'country': 'russia', 'city': 'factoria'}), tool_call_id='0', part_kind='tool-call')], timestamp=datetime.datetime(2024, 12, 18, 18, 53, 3, tzinfo=datetime.timezone.utc), kind='response'),
 ModelRequest(parts=[RetryPromptPart(content=[{'type': 'json_type', 'loc': (), 'msg': 'JSON input should be string, bytes or bytearray', 'input': {'country': 'russia', 'city': 'factoria'}}], tool_name='final_result', tool_call_id='0', timestamp=datetime.datetime(2024, 12, 18, 18, 53, 3, 498289, tzinfo=datetime.timezone.utc), part_kind='retry-prompt')], kind='request')]
```

**Comment by dmontagu** [comment]
It looks to me like it's expecting to get JSON, but it's getting an object. This might be an easy fix...

**Comment by dmontagu** [comment]
Okay, this is not an issue with pydantic-ai — the error is being raised internally in the openai SDK. In particular, the response from `TGI` appears to not be API-compatible with the openai SDK.

If I'm right about that, I expect that's not something we are going to try to resolve (TGI should just change their outputs to actually be OpenAI-compatible, if that's their goal), but I can imagine a couple ways forward if you need this to work in the short term:

* You can modify the implementation of OpenAIModel (or just implement a new TGIModel from scratch) to not use the OpenAI SDK, and instead use an SDK/manual HTTP requests compatible with TGI's tool call responses. (Or maybe there is a different version of the OpenAI SDK that is compatible with TGI? If that's the case, just switching to that might work.)
* You can pass a custom `client` to the `OpenAIModel` that will intercept the responses and transform them from the TGI-format to OpenAI format before handing them off to the OpenAI sdk.

**Comment by dmontagu** [comment]
Ah nevermind, I think the issue is the thing I thought it was at first, and the error response is a consequence of an earlier type error... I'll see if I can fix it..

**Comment by dmontagu** [comment]
It's hard for me to confirm as I'm not running TGI, but I believe #489 should fix these errors (and make the library more robust to passing messages produced by one model as inputs to another model).

I'm assuming the issue here is that the TGI response does not have the same types as the openai response, but I think it's worth being a bit more defensive against this particular API incompatibility. (I.e., receiving "loaded" JSON instead of a string).

**Comment by sadransh** [comment]
@dmontagu I did a quick test and I think the issue has been resolved with #489 . Thanks!
